src/kaleidoscope/hid.cpp

```diff
@@ -51,42 +52,172 @@
+// during the most recent cycle. For example, it would include modifiers attached to Key_A, but not 
+// modifiers attached to Key_LeftControl)
+
+static uint8_t requested_modifiers{0};
+
+// last_keycode_toggled_on is the keycode of the key most recently toggled on
+// for this report.  This is set when the key is pressed and cleared after
+// the report is sent.
+
+static uint8_t last_keycode_toggled_on{0};
+
+
+void resetModifierTracking(void) {
+  last_keycode_toggled_on = 0;
+  requested_modifiers = 0;
 }
 
-void pressRawKey(Key mappedKey) {
-  WITH_BOOTKEYBOARD_PROTOCOL {
-    BootKeyboard.press(mappedKey.keyCode);
-    return;
-  }
+// This returns true if the key is a keyboard key, and its keycode is a modifier.
+bool isModifierKey(Key mappedKey) {
+  // If it's not a keyboard key, return false
+  if (mappedKey.flags & (SYNTHETIC | RESERVED))
+    return false;
 
-  Keyboard.press(mappedKey.keyCode);
+  return (mappedKey.keyCode >= HID_KEYBOARD_FIRST_MODIFIER &&
+          mappedKey.keyCode <= HID_KEYBOARD_LAST_MODIFIER);
 }
 
-void pressKey(Key mappedKey) {
-  if (mappedKey.flags & SHIFT_HELD) {
+// requestModifiers takes a bitmap of modifiers that might apply
+// to the next USB HID report and adds them to a bitmap of all such modifiers.
+
+void requestModifiers(byte flags) {
+  requested_modifiers |= flags;
+}
+
+
+// pressModifiers takes a bitmap of modifier keys that must be included in
+// the upcoming USB HID report and passes them through to KeyboardioHID
+// immediately
+void pressModifiers(byte flags) {
+  if (flags & SHIFT_HELD) {
     pressRawKey(Key_LeftShift);
   }
-  if (mappedKey.flags & CTRL_HELD) {
+  if (flags & CTRL_HELD) {
     pressRawKey(Key_LeftControl);
   }
-  if (mappedKey.flags & LALT_HELD) {
+  if (flags & LALT_HELD) {
     pressRawKey(Key_LeftAlt);
   }
-  if (mappedKey.flags & RALT_HELD) {
+  if (flags & RALT_HELD) {
     pressRawKey(Key_RightAlt);
   }
-  if (mappedKey.flags & GUI_HELD) {
+  if (flags & GUI_HELD) {
     pressRawKey(Key_LeftGui);
   }
+}
+
+}
+
+void initializeKeyboard() {
+  Keyboard.begin();
+  WITH_BOOTKEYBOARD {
+    BootKeyboard.begin();
+  }
+}
+
+
+
+
+// pressKey takes a Key, as well as optional boolean 'toggledOn' which defaults to 'true'
+
+// If toggledOn is not set to false, this routine adds the modifier flags on this key
+// to the bitmask of modifiers that are allowed to be added to the upcoming report.
+// We do this so that when we roll over from a key with a modifier flag to one
+// without it, that modifier flag won't affect the new keypress.
+
+
+// If the key we're processing is a modifier key, any modifier flags attached to it are added
+// directly to the report along with the modifier from its keycode byte.
+//
+// (A 'modifier key' is one of the eight modifier keys defined by the HID standard:
+// left and right variants of Control, Shift, Alt, and GUI.)
+
+// Eventually it calls pressRawKey.
+
+void pressKey(Key mappedKey, boolean toggledOn) {
```

Comment:
The style guide specifies underscores for variable names, but since it matches the preexisting `mappedKey` I guess it makes sense to use `toggledOn` instead of `toggled_on`.

Reply:
Yup. Most important is that they should -match-, but that's not a reason that both of them shouldn't eventually change.